core/common/src/main/java/alluxio/Configuration.java

```diff
@@ -406,10 +407,36 @@ public static Duration getDuration(PropertyKey key) {
   }
 
   /**
-   * @return a view of the internal {@link Properties} of as an immutable map
+   * @return a map of the resolved properties represented by this configuration,
+   * including all default properties
    */
   public static Map<String, String> toMap() {
-    return Collections.unmodifiableMap(PROPERTIES);
+    Map<String, String> map = toRawMap();
+    // Resolve values with ${VALUE} format
+    String nestedPropIdentifier = "$";
+    for (Map.Entry<String, String> entry : map.entrySet()) {
+      String value = entry.getValue();
+      if (value != null && value.contains(nestedPropIdentifier)) {
+        map.put(entry.getKey(), lookup(value));
+      }
+    }
+    return map;
+  }
+
+  /**
+   * @return a map of the raw properties represented by this configuration,
+   * including all default properties
+   */
+  public static Map<String, String> toRawMap() {
+    Map<String, String> map = new HashMap<>(PROPERTIES);
+    for (PropertyKey key : PropertyKey.defaultKeys()) {
+      String propName = key.getName();
```

Comment:
maybe a simpler approach here and better code reuse?
```java
for (PropertyKey key : PropertyKey.defaultKeys()) {
  map.add(key.getName, get(key));
}
return map;
```

Reply:
lookupNonRecursively do not help a lot. It only helps that we do not need to check if the map.containsKey, but we still need to check if the defaultValue equals to null.

It is strange to use lookupNonRecursively here. We change key to keyName and then change it back to key inside lookupNonRecursively. When we check if the default value equals to null, we already getDefaultValue. lookupNonRecursively just getDefaultValue the second time.